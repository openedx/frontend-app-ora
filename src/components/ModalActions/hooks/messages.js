import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  finishLater: {
    id: 'ora-mfe.ModalActions.simpleAction.finishLater',
    defaultMessage: 'Save for later',
    description: 'Save for later (close) button text',
  },
  submitResponse: {
    id: 'ora-mfe.ModalActions.submitResponse',
    defaultMessage: 'Submit response',
    description: 'Submit button text',
  },
  submittingResponse: {
    id: 'ora-mfe.ModalActions.submittingResponse',
    defaultMessage: 'Submitting response',
    description: 'Submit button text while submitting',
  },
  responseSubmitted: {
    id: 'ora-mfe.ModalActions.responseSubmitted',
    defaultMessage: 'Response submitted',
    description: 'Submit button text after successful submission',
  },
  savingResponse: {
    id: 'ora-mfe.ModalActions.savingResponse',
    defaultMessage: 'Saving response',
    description: 'Save for later button text while saving',
  },
});

export default messages;
