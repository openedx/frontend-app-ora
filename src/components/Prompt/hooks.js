import { useState } from 'react';

const usePromptHooks = () => {
  const [open, setOpen] = useState(true);

  const toggleOpen = () => setOpen(!open);

  return {
    open,
    toggleOpen,
  };
};

export default usePromptHooks;
