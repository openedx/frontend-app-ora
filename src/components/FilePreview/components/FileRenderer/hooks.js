import { useState } from 'react';
import { errorStatuses, errorMessages, renderers } from '../constants';
import { getFileType } from '../utils';
import messages from './messages';

/**
 * component hooks
 */
export const useRenderData = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useState(null);
  const [isLoading, setIsLoading] = useState(true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    errorMessages[errorStatus] || errorMessages[errorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = renderers[getFileType(file.fileName)];
  const rendererProps = {
    fileName: file.fileName,
    url: file.fileUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
