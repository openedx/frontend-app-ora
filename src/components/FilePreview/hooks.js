import { useKeyedState, StrictDict } from '@edx/react-unit-test-utils';

import {
  PDFRenderer,
  ImageRenderer,
  TXTRenderer,
} from 'components/FilePreview/BaseRenderers';

import messages from './messages';

export const ErrorStatuses = StrictDict({
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  conflict: 409,
  serverError: 500,
});

export const FileTypes = StrictDict({
  pdf: 'pdf',
  jpg: 'jpg',
  jpeg: 'jpeg',
  png: 'png',
  bmp: 'bmp',
  txt: 'txt',
  gif: 'gif',
  jfif: 'jfif',
  pjpeg: 'pjpeg',
  pjp: 'pjp',
  svg: 'svg',
});

/**
 * Config data
 */
export const RENDERERS = {
  [FileTypes.pdf]: PDFRenderer,
  [FileTypes.jpg]: ImageRenderer,
  [FileTypes.jpeg]: ImageRenderer,
  [FileTypes.bmp]: ImageRenderer,
  [FileTypes.png]: ImageRenderer,
  [FileTypes.txt]: TXTRenderer,
  [FileTypes.gif]: ImageRenderer,
  [FileTypes.jfif]: ImageRenderer,
  [FileTypes.pjpeg]: ImageRenderer,
  [FileTypes.pjp]: ImageRenderer,
  [FileTypes.svg]: ImageRenderer,
};

export const SUPPORTED_TYPES = Object.keys(RENDERERS);

export const ERROR_STATUSES = {
  [ErrorStatuses.notFound]: messages.fileNotFoundError,
  [ErrorStatuses.serverError]: messages.unknownError,
};

/**
 * Util methods and transforms
 */
export const getFileType = (fileName) => fileName.split('.').pop()?.toLowerCase();
export const isSupported = (file) => SUPPORTED_TYPES.includes(
  getFileType(file.name),
);

export const stateKeys = StrictDict({
  errorStatus: 'errorStatus',
  isLoading: 'isLoading',
});

/**
 * component hooks
 */
export const renderHooks = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useKeyedState(stateKeys.errorStatus, null);
  const [isLoading, setIsLoading] = useKeyedState(stateKeys.isLoading, true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    ERROR_STATUSES[errorStatus] || ERROR_STATUSES[ErrorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = RENDERERS[getFileType(file.name)];
  const rendererProps = {
    fileName: file.name,
    url: file.downloadUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
