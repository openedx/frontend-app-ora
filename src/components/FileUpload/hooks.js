import { useCallback } from 'react';

import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';
import { useDownloadFiles } from 'hooks/app';

export const stateKeys = StrictDict({
  shouldShowError: 'shouldShowError',
  isModalOpen: 'isModalOpen',
  uploadArgs: 'uploadArgs',
  description: 'description',
});

export const useUploadConfirmModalHooks = ({
  file,
  closeHandler,
  uploadHandler,
}) => {
  const [description, setDescription] = useKeyedState(stateKeys.description, '');
  const [shouldShowError, setShouldShowError] = useKeyedState(
    stateKeys.shouldShowError,
    false,
  );

  const confirmUploadClickHandler = () => {
    // console.log({ confirmUploadClick: { description } });
    if (description !== '') {
      uploadHandler(file, description);
    } else {
      setShouldShowError(true);
    }
  };

  const exitHandler = () => {
    setShouldShowError(false);
    setDescription('');
    closeHandler();
  };

  // Modifying pointer of file object. This is not a good practice.
  // eslint-disable-next-line no-param-reassign, no-return-assign
  const onFileDescriptionChange = (event) => setDescription(event.target.value);

  return {
    shouldShowError,
    confirmUploadClickHandler,
    exitHandler,
    onFileDescriptionChange,
  };
};

export const useFileUploadHooks = ({ onFileUploaded }) => {
  const [uploadArgs, setUploadArgs] = useKeyedState(stateKeys.uploadArgs, {});
  const [isModalOpen, setIsModalOpen] = useKeyedState(
    stateKeys.isModalOpen,
    false,
  );

  const confirmUpload = useCallback(async (file, description) => {
    // console.log({ confirmUpload: { file, description } });
    setIsModalOpen(false);
    if (onFileUploaded) {
      // console.log({ uploadArgs });
      await onFileUploaded({ ...uploadArgs, description });
    }
    setUploadArgs({});
  }, [uploadArgs, onFileUploaded, setIsModalOpen, setUploadArgs]);

  const closeUploadModal = useCallback(() => {
    setIsModalOpen(false);
    setUploadArgs({});
  }, [setIsModalOpen, setUploadArgs]);

  const onProcessUpload = useCallback(
    ({ fileData, handleError, requestConfig }) => {
      setIsModalOpen(true);
      setUploadArgs({ fileData, handleError, requestConfig });
    },
    [setIsModalOpen, setUploadArgs],
  );

  return {
    isModalOpen,
    uploadArgs,
    confirmUpload,
    closeUploadModal,
    onProcessUpload,
  };
};

export const useFileDownloadHooks = ({ files, zipFileName }) => {
  const downloadFileMation = useDownloadFiles();
  const downloadFiles = () => downloadFileMation.mutate({
    files,
    zipFileName,
  });
  return {
    downloadFiles,
    status: downloadFileMation.status,
  };
};

export default {
  useUploadConfirmModalHooks,
  useFileUploadHooks,
  useFileDownloadHooks,
};
