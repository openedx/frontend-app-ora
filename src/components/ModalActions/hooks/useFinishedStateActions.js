import { useGlobalState, useTrainingStepIsCompleted } from 'hooks/app';
import {
  useHasSubmitted,
  useSubmittedAssessment,
} from 'hooks/assessment';
import { useViewStep } from 'hooks/routing';
import {
  useStartStepAction,
  useLoadNextAction,
  useCloseModalAction,
} from 'hooks/actions';
import { stepNames, stepStates } from 'constants/index';

const useFinishedStateActions = () => {
  const step = useViewStep();
  const globalState = useGlobalState({ step });
  const hasSubmitted = useHasSubmitted();
  const startStepAction = useStartStepAction(step);
  const submittedAssessment = useSubmittedAssessment();
  const loadNextAction = useLoadNextAction();
  const trainingStepIsCompleted = useTrainingStepIsCompleted();

  const stepState = globalState.activeStepState;

  const exitAction = useCloseModalAction();

  if (!hasSubmitted) {
    if (step === stepNames.studentTraining && trainingStepIsCompleted) {
      return { primary: startStepAction, secondary: exitAction };
    }
    return null;
  }

  // assessment finished state
  if (submittedAssessment) {
    const { activeStepName } = globalState;
    if (activeStepName === stepNames.done) {
      return { primary: startStepAction };
    }
    if (activeStepName === stepNames.staff) {
      return { primary: exitAction };
    }
    // finished and moved to next step
    if ([stepNames.submission || stepNames.self].includes(step)) {
      return { primary: startStepAction, secondary: exitAction };
    }
    if (step !== activeStepName) {
      // next step is available
      if (stepState === stepStates.inProgress) {
        return { primary: startStepAction, secondary: exitAction };
      }
      // next step is not available
      return null;
    }

    // finished current assessment but not current step
    if (stepState === stepStates.inProgress) {
      return { primary: loadNextAction, secondary: exitAction };
    }
    // finished current assessment, but not step
    // and there are no more assessments available for the current step
    return { primary: exitAction };
  }
  // submission finished state
  return { primary: startStepAction, secondary: exitAction };
};

export default useFinishedStateActions;
