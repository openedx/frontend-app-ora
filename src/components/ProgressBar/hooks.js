import { useParams } from 'react-router-dom';
import { useIsEmbedded, useViewStep } from 'hooks';
import {
  useEffectiveGrade,
  useGlobalState,
} from 'data/services/lms/hooks/selectors';
import {
  stepRoutes,
  stepStates,
} from 'data/services/lms/constants';

export const useProgressStepData = ({ step, canRevisit = false }) => {
  const { xblockId, courseId } = useParams();
  const isEmbedded = useIsEmbedded();
  const viewStep = useViewStep();
  const { stepState } = useGlobalState({ step });

  const href = `/${stepRoutes[step]}${isEmbedded ? '/embedded' : ''}/${courseId}/${xblockId}`;
  const isActive = viewStep === step;
  const isEnabled = (
    isActive
    || (stepState === stepStates.inProgress)
    || (canRevisit && stepState === stepStates.done)
  );
  const myGrade = useEffectiveGrade()?.stepScore;
  return {
    href,
    isEnabled,
    isActive,
    isComplete: stepState === stepStates.done,
    inProgress: stepState === stepStates.inProgress,
    isPastDue: stepState === stepStates.closed,
    myGrade,
    // myGrade: { earned: 8, possible: 10 },
    // isPastDue: step === 'self',
  };
};

export default { useProgressStepData };
