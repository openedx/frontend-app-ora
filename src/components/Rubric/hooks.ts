import { useKeyedState, StrictDict } from '@edx/react-unit-test-utils';
import { usePageData, useRubricConfig } from 'data/services/lms/hooks/selectors';
import { submitRubric } from 'data/services/lms/hooks/actions';
import { RubricHookData } from './types';

export const stateKeys = StrictDict({
  rubric: 'rubric',
  overallFeedback: 'overallFeedback',
});

export const useRubricData = ({ isGrading }): RubricHookData => {
  const data = usePageData();
  const { criteria, feedbackConfig } = useRubricConfig();

  const [rubricData, setRubricData] = useKeyedState(stateKeys.rubric, data.rubric);
  const [overallFeedback, setOverallFeedback] = useKeyedState(
    stateKeys.overallFeedback, data.rubric.overallFeedback
  );
  const submitRubricMutation = submitRubric();

  const onOverallFeedbackChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    setOverallFeedback(e.target.value);

  const onSubmit = () => {
    submitRubricMutation.mutate({
      overallFeedback,
      optionsSelected: rubricData.optionsSelected,
      criterionFeedback: rubricData.criterionFeedback,
    } as any);
  };

  return {
    rubricData,
    setRubricData,
    criteria: criteria.map((criterion) => ({
      ...criterion,
      optionsValue: rubricData.optionsSelected[criterion.name],
      optionsIsInvalid: false,
      optionsOnChange: (e: React.ChangeEvent<HTMLInputElement>) => {
        setRubricData({
          ...rubricData,
          optionsSelected: {
            ...rubricData.optionsSelected,
            [criterion.name]: e.target.value,
          },
        });
      },

      feedbackValue: rubricData.criterionFeedback[criterion.name],
      feedbackIsInvalid: false,
      feedbackOnChange: (e: React.ChangeEvent<HTMLInputElement>) => {
        setRubricData({
          ...rubricData,
          criterionFeedback: {
            ...rubricData.criterionFeedback,
            [criterion.name]: e.target.value,
          },
        });
      },
    })),
    onSubmit,
    submitStatus: submitRubricMutation.status,

    // overall feedback
    overallFeedback,
    onOverallFeedbackChange,
    overallFeedbackDisabled: !isGrading,
    overallFeedbackIsInvalid: false,
    overallFeedbackPrompt: feedbackConfig.defaultText,
  };
};

export default useRubricData;
