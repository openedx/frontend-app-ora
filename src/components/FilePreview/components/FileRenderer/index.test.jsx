import { shallow } from '@edx/react-unit-test-utils';

import { useRenderData } from './hooks';
import { FileRenderer } from './index';

jest.mock('./Banners', () => ({
  ErrorBanner: () => 'ErrorBanner',
  LoadingBanner: () => 'LoadingBanner',
}));

jest.mock('./hooks', () => ({
  useRenderData: jest.fn(),
}));

describe('FileRenderer Component', () => {
  const props = {
    file: {
      fileName: 'some_file',
      fileUrl: 'some_url',
    },
    defaultOpen: true,
  };

  const defaultRenderData = {
    Renderer: () => 'Renderer',
    isLoading: false,
    errorStatus: false,
    error: null,
    rendererProps: {
      abc: 123,
    },
  };

  it('render default', () => {
    useRenderData.mockReturnValue(defaultRenderData);
    const wrapper = shallow(<FileRenderer {...props} />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('Renderer')).toHaveLength(1);
    expect(wrapper.instance.findByType('LoadingBanner')).toHaveLength(0);
    expect(wrapper.instance.findByType('ErrorBanner')).toHaveLength(0);
  });

  it('render loading banner', () => {
    useRenderData.mockReturnValue({ ...defaultRenderData, isLoading: true });
    const wrapper = shallow(<FileRenderer {...props} />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('LoadingBanner')).toHaveLength(1);
    expect(wrapper.instance.findByType('ErrorBanner')).toHaveLength(0);
    expect(wrapper.instance.findByType('Renderer')).toHaveLength(0);
  });

  it('render error banner', () => {
    useRenderData.mockReturnValue({ errorStatus: true, error: { message: 'some_error' } });
    const wrapper = shallow(<FileRenderer {...props} />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('ErrorBanner')).toHaveLength(1);
    expect(wrapper.instance.findByType('LoadingBanner')).toHaveLength(0);
  });
});
