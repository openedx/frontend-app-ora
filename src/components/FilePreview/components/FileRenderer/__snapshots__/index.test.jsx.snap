// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileRenderer Component render default 1`] = `
<FileCard
  defaultOpen={true}
  file={
    Object {
      "fileName": "some_file",
      "fileUrl": "some_url",
    }
  }
>
  <Renderer
    abc={123}
  />
</FileCard>
`;

exports[`FileRenderer Component render error banner 1`] = `
<FileCard
  defaultOpen={true}
  file={
    Object {
      "fileName": "some_file",
      "fileUrl": "some_url",
    }
  }
>
  <ErrorBanner
    message="some_error"
  />
</FileCard>
`;

exports[`FileRenderer Component render loading banner 1`] = `
<FileCard
  defaultOpen={true}
  file={
    Object {
      "fileName": "some_file",
      "fileUrl": "some_url",
    }
  }
>
  <LoadingBanner />
</FileCard>
`;
